### ISSUE TITLE
[Performance] Implement Real Batch Processing in AutoregressiveGenerator - Remove Fallback to Single Token Generation

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#280]
# Problem Description

The `BatchProcessor` struct in `AutoregressiveGenerator` is currently a placeholder implementation. The `generate_token_batched` method falls back to single token generation instead of implementing true batch processing, significantly limiting inference throughput for concurrent requests.

## Environment

- **File**: `crates/bitnet-inference/src/generation/autoregressive.rs`
- **Struct**: `BatchProcessor` (lines 152-166)
- **Method**: `AutoregressiveGenerator::generate_token_batched` (lines 383-396)
- **Rust Version**: 1.90.0 (2024 edition)
- **Features**: Affects both `cpu` and `gpu` feature builds

## Current Implementation

```rust
async fn generate_token_batched<F, Fut>(
    &mut self,
    current_tokens: &[usize],
    forward_fn: &F,
    step: usize,
) -> Result<usize>
where
    F: Fn(BitNetTensor) -> Fut + Send + Sync,
    Fut: std::future::Future<Output = Result<BitNetTensor>> + Send,
{
    // For now, fallback to single token generation
    // In a full implementation, this would batch multiple sequences
    self.generate_token_single(current_tokens, forward_fn, step).await
}
```

### Current State
1. **Stub Implementation**: `BatchProcessor` struct exists but contains only placeholder fields marked with `#[allow(dead_code)]`
2. **Fallback Behavior**: `generate_token_batched` immediately delegates to `generate_token_single`
3. **Performance Mode Ignorance**: Even when `PerformanceMode::Throughput` is selected, batching is not utilized
4. **Resource Waste**: GPU/CPU resources are underutilized as inference operates on single sequences

### Affected Components
- **AutoregressiveGenerator**: Primary inference engine
- **BitNet Inference Server**: Production throughput capabilities
- **Concurrent Request Handling**: Multiple client scenarios
- **GPU Utilization**: Batch-friendly CUDA kernels remain underutilized

### Performance Impact
- **Throughput Loss**: 3-10x lower inference throughput compared to proper batching
- **GPU Underutilization**: CUDA cores remain idle during single-sequence inference
- **Latency Inefficiency**: Higher per-request latency due to lack of batching optimizations
- **Resource Waste**: Memory allocation overhead for individual requests

### Business Impact
- **Production Readiness**: Server cannot efficiently handle concurrent inference requests
- **Scalability Limitations**: Poor horizontal scaling characteristics
- **Cost Inefficiency**: Higher infrastructure costs due to underutilized hardware

### Affected Use Cases
- Multi-user inference servers
- Batch inference workloads  
- Production deployment scenarios
- Performance benchmarking

### Core Implementation Strategy

Replace the stub `generate_token_batched` with a real batching implementation:

```rust
async fn generate_token_batched<F, Fut>(
    &mut self,
    batch_requests: &[BatchRequest],
    forward_fn: &F,
) -> Result<Vec<usize>>
where
    F: Fn(BitNetTensor) -> Fut + Send + Sync,
    Fut: std::future::Future<Output = Result<BitNetTensor>> + Send,
{
    // 1. Collect and pad input sequences to same length
    let (batched_tensor, padding_info) = self.prepare_batch_tensor(batch_requests)?;
    
    // 2. Single forward pass for entire batch
    let batch_logits = forward_fn(batched_tensor).await?;
    
    // 3. Extract and distribute results back to individual requests
    let tokens = self.extract_batch_results(batch_logits, padding_info).await?;
    
    Ok(tokens)
}
```

#### Phase 1: Core Batch Processing (High Priority)
1. **Implement `BatchRequest` queue management**
   - Add request collection mechanism
   - Implement batch size optimization
   - Add timeout-based batch flushing

2. **Develop tensor batching utilities**
   - Sequence padding for uniform length
   - Attention mask generation
   - Memory-efficient tensor concatenation

3. **Create batch result distribution**
   - Extract per-sequence results from batch output
   - Handle variable-length output sequences
   - Maintain request-to-result mapping

#### Phase 2: Performance Optimization (Medium Priority)
1. **Adaptive batch sizing**
   - Dynamic batch size based on memory availability
   - Latency vs throughput trade-off optimization
   - Device-specific batch size tuning

2. **Memory optimization**
   - Efficient padding strategies
   - Tensor memory pooling
   - Gradient-free inference optimization

#### Phase 3: Advanced Features (Lower Priority)
1. **Smart request scheduling**
   - Priority-based batching
   - Similar-length sequence grouping
   - Deadline-aware scheduling

2. **Cross-device optimization**
   - CPU/GPU hybrid batching
   - Memory transfer optimization
   - Pipeline parallelism

#### Data Structures
```rust
#[derive(Debug)]
pub struct BatchRequest {
    pub input_tokens: Vec<usize>,
    pub max_new_tokens: usize,
    pub generation_config: GenerationConfig,
    pub request_id: String,
    pub priority: u8,
    pub deadline: Option<Instant>,
}

#[derive(Debug)]
pub struct BatchProcessor {
    pending_requests: VecDeque<BatchRequest>,
    max_batch_size: usize,
    batch_timeout_ms: u64,
    memory_budget: usize,
}
```

#### Integration Points
- **Autoregressive Generator**: Main inference loop integration
- **Performance Modes**: Throughput mode optimization
- **Memory Management**: Tensor cache coordination
- **Error Handling**: Batch failure recovery

#### Unit Tests
- Batch tensor preparation correctness
- Result extraction accuracy  
- Memory usage validation
- Edge case handling (empty batches, single requests)

#### Integration Tests
- End-to-end batch inference pipeline
- Performance mode switching
- Concurrent request handling
- Memory pressure scenarios

#### Performance Tests
- Throughput measurement (tokens/second)
- Latency distribution analysis
- Memory usage profiling
- GPU utilization monitoring

#### Validation Requirements
- Cross-validation with single-token generation results
- Numerical accuracy verification
- Memory leak detection
- Resource cleanup validation

### Functional Requirements
- [ ] `generate_token_batched` processes multiple sequences in single forward pass
- [ ] Batch processing maintains identical results to single-token generation
- [ ] Adaptive batch sizing based on available memory and performance mode
- [ ] Proper error handling and recovery for batch failures
- [ ] Request queuing and timeout-based batch formation

### Performance Requirements  
- [ ] 3-5x throughput improvement for batched inference (batch size ≥ 4)
- [ ] Memory usage scales linearly with batch size
- [ ] GPU utilization >80% during batch processing
- [ ] Latency overhead <10% compared to optimal single-request latency

### Integration Requirements
- [ ] Seamless integration with existing `AutoregressiveGenerator` API
- [ ] Compatible with all performance modes (Latency, Throughput, Balanced, Conservative)
- [ ] Works with both CPU and GPU feature builds
- [ ] Maintains deterministic generation when `BITNET_DETERMINISTIC=1`

### Quality Requirements
- [ ] Comprehensive test coverage (>90% for new batch processing code)
- [ ] No memory leaks under sustained batch processing
- [ ] Graceful degradation when batch processing fails
- [ ] Clear logging and monitoring for batch operations

## Related Issues

This issue is part of a broader production readiness initiative:

- **Stub Implementations**: Multiple other stub implementations need completion
- **Performance Optimization**: GPU memory management optimizations
- **Production Server**: Batch processing is critical for production inference server
- **Monitoring**: Batch processing metrics for observability

### Memory Considerations
- Implement memory pressure detection to prevent OOM
- Use efficient padding strategies (left-padding for causal models)
- Consider memory pool allocation for batch tensors

### Device Compatibility
- Ensure CUDA kernel compatibility with batched operations
- Optimize for different GPU architectures (V100, A100, RTX series)
- Maintain CPU fallback for smaller batches

### Error Recovery
- Implement individual request retry on batch failure
- Graceful degradation to single-token generation
- Clear error attribution for debugging

### Monitoring Integration
- Expose batch processing metrics
- Track batch formation efficiency
- Monitor memory usage patterns

---

**Priority**: High
**Complexity**: Medium-High  
**Estimated Effort**: 2-3 weeks
**Dependencies**: None (can be implemented independently)
**Breaking Changes**: None (backward compatible)
